Pinpoint toJSON methods are being generated incorrectly

Hi, we've had our own internal fork of the old Amazonka. It had a bug where the toJSON for various calls were incorrect - it's adding an extra key when it should be flat). It looks like 2.0 has the same issue.

For example:
* https://github.com/brendanhay/amazonka/blob/29e4888f847797533c968a855804990ecf8660dd/lib/services/amazonka-pinpoint/gen/Amazonka/Pinpoint/UpdateEndpoint.hs#L147
* https://github.com/brendanhay/amazonka/blob/29e4888f847797533c968a855804990ecf8660dd/lib/services/amazonka-pinpoint/gen/Amazonka/Pinpoint/UpdateSegment.hs#L145
* https://github.com/brendanhay/amazonka/blob/29e4888f847797533c968a855804990ecf8660dd/lib/services/amazonka-pinpoint/gen/Amazonka/Pinpoint/SendMessages.hs#L124

```instance Core.ToJSON UpdateEndpoint where
  toJSON UpdateEndpoint' {..} =
    Core.object
      ( Prelude.catMaybes
          [ Prelude.Just
              ("EndpointRequest" Core..= endpointRequest)
          ]
      )
```

The "EndpointRequest" should be removed (something like):

```
instance Core.ToJSON UpdateEndpoint where
  toJSON UpdateEndpoint' {..} = Core.toJSON endpointRequest
```

The error you get back from AWS (the pertinent bit is "Unrecognized field "EndpointRequest"):
```
Left (ServiceError (ServiceError' {_serviceErrorAbbrev = Abbrev {fromAbbrev = "Pinpoint"}, _serviceErrorStatus = Status {statusCode = 400, statusMessage = "Bad Request"}, _serviceErrorHeaders = [("Content-Type","application/json"),("Content-Length","260"),("Connection","keep-alive"),("Date","Tue, 05 Apr 2022 04:36:40 GMT"),("x-amzn-RequestId","21d60202-29e8-4df2-8f0f-2ed5f7504e95"),("Access-Control-Allow-Origin","*"),("x-amzn-ErrorType","BadRequestException"),("x-amz-apigw-id","QFwV0G-3SwMFXNg="),("Cache-Control","no-store"),("X-Amzn-Trace-Id","Root=1-624bc758-05e2608a66a7952a043beb65"),("X-Cache","Error from cloudfront"),("Via","1.1 4cc1cef5b7e8e0c2a1020c88ce286fbc.cloudfront.net (CloudFront)"),("X-Amz-Cf-Pop","SYD4-C1"),("X-Amz-Cf-Id","OjLJ0IXAhU_DZywSM7-7V1FaAlzVWCzNrfNSDGJdHnDN3ELG7TTwkQ==")], _serviceErrorCode = ErrorCode "BadRequest", _serviceErrorMessage = Just (ErrorMessage {fromErrorMessage = "Unrecognized field \"EndpointRequest\" (11 known properties: \"Metrics\", \"Address\", \"Location\", \"EndpointStatus\", \"RequestId\", \"OptOut\", \"ChannelType\", \"Demographic\", \"Attributes\", \"User\", \"EffectiveDate\"])\nLine 1, column 21"}), _serviceErrorRequestId = Just (RequestId {fromRequestId = "21d60202-29e8-4df2-8f0f-2ed5f7504e95"})}))
```

You can see how they're supposed to be structured:

* [Endpoint Schemas](https://docs.aws.amazon.com/pinpoint/latest/apireference/apps-application-id-endpoints-endpoint-id.html#apps-application-id-endpoints-endpoint-id-schemas)
* [Segment Schemas](https://docs.aws.amazon.com/pinpoint/latest/apireference/apps-application-id-segments-segment-id.html#apps-application-id-segments-segment-id-schemas)
* [Messages Schema](https://docs.aws.amazon.com/pinpoint/latest/apireference/apps-application-id-messages.html#apps-application-id-messages-schemas)

Hi, I gave it a go and haven't got back to it I got this compiler error ```amazonka-pinpoint             ../amazonka-pinpoint-2.0/gen/Amazonka/Pinpoint/TagResource.hs:113:30: error:
amazonka-pinpoint             >     Not in scope: data constructor ‘Core.ToJSON’
amazonka-pinpoint             >     Perhaps you meant variable ‘Core.toJSON’ (imported from Amazonka.Core)
amazonka-pinpoint             >     Module ‘Amazonka.Core’ does not export ‘ToJSON’.
amazonka-pinpoint             >     |
amazonka-pinpoint             > 113 |   toJSON TagResource' {..} = Core.ToJSON tagsModel
amazonka-pinpoint             >     |                              ^^^^^^^^^^^
amazonka-pinpoint             >```

--- a/gen/src/Gen/AST/Data/Syntax.hs
+++ b/gen/src/Gen/AST/Data/Syntax.hs
@@ -7,6 +7,7 @@ import qualified Control.Lens as Lens
 import qualified Data.Char as Char
 import qualified Data.Foldable as Fold
 import qualified Data.HashMap.Strict as HashMap
+import Data.List (find)
 import qualified Data.List.NonEmpty as NE
 import qualified Data.Text as Text
 import Gen.AST.Data.Field
@@ -524,16 +525,19 @@ toXMLD p n =
     enc = mconcatE . map (either id id)
 
 toJSOND :: Protocol -> Id -> [Field] -> Decl
-toJSOND p n =
+toJSOND p n fs =
   instD1 "Core.ToJSON" n
     . wildcardD n "toJSON" enc (Exts.paren $ Exts.app (var "Core.Object") memptyE)
-    . map (Right . toJSONE p)
+    $ map (Right . toJSONE p) fs
   where
-    enc =
-      Exts.app (var "Core.object")
-        . Exts.app (var "Prelude.catMaybes")
-        . Exts.listE
-        . map (either id id)
+    enc = case find _fieldPayload fs of
+      Nothing ->
+        Exts.app (var "Core.object")
+          . Exts.app (var "Prelude.catMaybes")
+          . Exts.listE
+          . map (either id id)
+      Just f ->
+        const $ Exts.app (var "Core.toJSON") (var $ fieldAccessor f)
 
 toHeadersD :: Protocol -> Id -> [Either (Text, Text) Field] -> Decl
 toHeadersD p n = instD1 "Core.ToHeaders" n . wildcardD n "toHeaders" enc memptyE
@@ -572,7 +576,7 @@ wildcardD ::
 wildcardD n f enc xs = \case
   [] -> constD f xs
   es
-    | not (any isRight es) -> funD f $ Exts.app (var "Prelude.const") (enc es)
+    | all isLeft es -> funD f $ Exts.app (var "Prelude.const") (enc es)
     | otherwise -> Exts.InsDecl () (Exts.FunBind () [match prec es])
   where
     match p es =

--- a/lib/amazonka/CHANGELOG.md
+++ b/lib/amazonka/CHANGELOG.md
@@ -53,6 +53,9 @@ Released: **?**, Compare: [2.0.0-rc1](https://github.com/brendanhay/amazonka/com
 
 ### Fixed
 
+- Generator: Correctly generate `ToJSON` instances for requests which set a `"payload":` field in `"type": "structure"` definitions.
+[\#790](https://github.com/brendanhay/amazonka/pull/790)
+  - Fixes some API calls in `amazonka-glacier` and `amazonka-pinpoint`
 - Presigning URLs that are not for S3
 [\#767](https://github.com/brendanhay/amazonka/pull/767)
 - `amazonka-s3`/`amazonka-glacier`: treat upload IDs are a mandatory part of the `CreateMultipartUpload`/`InitiateMultipartUpload` responses.

